packages/datadog-esbuild/index.js

```diff
@@ -0,0 +1,104 @@
+'use strict'
+
+/* eslint-disable no-console */
+
+const NAMESPACE = 'datadog'
+
+const instrumented = Object.keys(require('../datadog-instrumentations/src/helpers/hooks.js'))
+const rawBuiltins = require('module').builtinModules
+
+warnIfUnsupported()
+
+const builtins = new Set()
+
+for (const builtin of rawBuiltins) {
+  builtins.add(builtin)
+  builtins.add(`node:${builtin}`)
+}
+
+const packagesOfInterest = new Set()
+
+const DEBUG = !!process.env.DD_TRACE_DEBUG
+
+// We don't want to handle any built-in packages via DCITM
+// Those packages will still be handled via RITM
+// Attempting to instrument them would fail as they have no package.json file
+for (const pkg of instrumented) {
+  if (builtins.has(pkg)) continue
+  if (pkg.startsWith('node:')) continue
+  packagesOfInterest.add(pkg)
+}
+
+const DC_CHANNEL = 'dd-trace:bundledModuleLoadStart'
+
+module.exports.name = 'datadog-esbuild'
+
+module.exports.setup = function (build) {
+  build.onResolve({ filter: /.*/ }, args => {
+    const packageName = args.path
+
+    if (args.namespace === 'file' && packagesOfInterest.has(packageName)) {
+      // The file namespace is used when requiring files from disk in userland
+      const pathToPackageJson = require.resolve(`${packageName}/package.json`, { paths: [ args.resolveDir ] })
+      const pkg = require(pathToPackageJson)
+
+      if (DEBUG) {
+        console.log(`resolve ${packageName}@${pkg.version}`)
+      }
+
+      // https://esbuild.github.io/plugins/#on-resolve-arguments
+      return {
+        path: packageName,
+        namespace: NAMESPACE,
+        pluginData: {
+          version: pkg.version
+        }
+      }
+    } else if (args.namespace === 'datadog') {
+      // The datadog namespace is used when requiring files that are injected during the onLoad stage
+      // see note in onLoad
+
+      if (builtins.has(packageName)) return
+
+      return {
+        path: require.resolve(packageName, { paths: [ args.resolveDir ] }),
+        namespace: 'file'
+      }
+    }
+  })
+
+  build.onLoad({ filter: /.*/, namespace: NAMESPACE }, args => {
+    if (DEBUG) {
+      console.log(`load ${args.path}@${args.pluginData.version}`)
+    }
+
+    // JSON.stringify adds double quotes. For perf gain could simply add in quotes when we know it's safe.
+    const contents = `
+      const dc = require('diagnostics_channel');
+      const ch = dc.channel(${JSON.stringify(DC_CHANNEL + ':' + args.path)});
```

Comment:
Any specific reason to make that per plugin? It requires subscribers to know in advance what modules to subscribe to, which might not be the case when this supports `import` as well.

Reply:
I spoke with Stephen about this regarding performance and he suggested there wasn't a big difference between the two approaches.

With the per-library-channel approach the ability to [subscribe and trigger the appropriate callback](https://github.com/DataDog/dd-trace-js/pull/2763/files#diff-c493ba28bea854a0366a4ffcd13abf71e2c36cb0a7d6a2c0711a825678de42f3R40) is simpler.

With a single channel approach I think I would need to use the registry pattern with a singleton listener on the channel that dispatches requests to registered callbacks stored in a singleton registry. I could be wrong though and there might be a cleaner approach.

I'm curious, how might this be more difficult using `import` vs `require`? Wouldn't we know the name of npm packages ahead of time with either loader?